### DISCUSSION TITLE
KCL migration

### DISCUSSION THREAD

**Discussion by Jip-Hop** [zbm-dev/zfsbootmenu#287] [Questions]
I wrote a guide to install Openmediavault with a ZFS root filesystem. The guide uses a fun trick: it (ab)uses Proxmox to install Openmediavault. At [step 5 of the guide](https://github.com/Jip-Hop/OMV-ZFS-RAID-NAS/tree/main#step-5-booting-omv-and-cleanup) I'm presented with a warning by ZFSBootMenu:

```
Using KCL from /etc/default/grub on pool rpool/omv
This behavior is DEPRECATED and will be removed soon

KCL should be migrated to an org.zfsbootmenu:commandline property.

Will attempt migration in 60 seconds
```

What can I do to prevent this issue? Is there anything to worry about now, or potentially in the future when this depricated behavior is removed?

Do I actually need the `KCL` from `/etc/default/grub` in order to boot?

Thanks a lot for making ZFSBootMenu by the way. It's really awesome!

**Comment by zdykstra** [comment]
Unless you've actually set a custom kernel commandline in `/etc/default/grub`, you don't need it. Simply remove the file to prevent the migration prompt from happening. If you do need to set custom values, set `org.zfsbootmenu:commandline` on the boot environment or the pool.

Briefly reading through your guide, I see that you're having people run `zpool import -af` . ZFSBootMenu has gone to great lengths to avoid ever needing to do something as drastic/dangerous as force importing all pools. Can you share the errors that you get when you run `zpool import -a` on first boot?

**Comment by Jip-Hop** [comment]
Thanks for replying :)

```
zpool import -a
cannot import 'rpool': pool was previously in use from another system.
Last accessed by (none) (hostid=f6568c23) at date + time
The pool can be imported, use 'zpool import -f' to import the pool.
```

Tried once more, but this time I ran `export rpool` before rebooting into ZFSBootMenu.

```
zpool import -a
cannot import 'rpool': pool was previously in use from another system.
Last accessed by (none) (hostid=a6c334a0) at date + time
The pool can be imported, use 'zpool import -f' to import the pool.
```

Instead of `zpool import -af` I could do: `zpool import -f rpool`. But would be nice if I didn't have to manually import at all.

I renamed `/etc/default/grub` and didn't get the KCL migration warning. The system still boots. Running `update-grub` completed successfully and didn't re-introduce the warning (that's important I think, as the installed system may run this occasionally as it's not 'aware' I'm using ZFSBootMenu).

**Reply by ahesford** [reply]
You should boot ZFSBootMenu with `zbm.import_policy=hostid` to avoid these import failures. Alternatively, boot ZBM with `spl_hostid=<whatever-you-need>` to match what is recorded in the pools. We also recommend `zbm.set_hostid` to force the value in the boot environment.

You don't need to rename the GRUB configuration if you set the `org.zfsbootmenu:commandline` property.

Why is grub installed in the first place?
